crates/wast/src/component/func.rs

```diff
@@ -460,6 +556,602 @@ impl Default for CanonThreadHwConcurrency {
     }
 }
 
+/// Information relating to the `task.return` intrinsic.
+#[derive(Debug)]
+pub struct CanonTaskReturn<'a> {
+    /// The core function type representing the signature of this intrinsic.
+    pub ty: Index<'a>,
+}
+
+impl<'a> Parse<'a> for CanonTaskReturn<'a> {
+    fn parse(parser: Parser<'a>) -> Result<Self> {
+        parser.parse::<kw::task_return>()?;
+
+        Ok(Self {
+            ty: parser.parse()?,
+        })
+    }
+}
+
+impl Default for CanonTaskReturn<'_> {
+    fn default() -> Self {
+        Self {
+            ty: Index::Num(0, Span::from_offset(0)),
+        }
+    }
+}
+
+/// Information relating to the `task.wait` intrinsic.
+#[derive(Debug)]
+pub struct CanonTaskWait<'a> {
+    /// If true, the component instance may be reentered during a call to this
+    /// intrinsic.
+    pub async_: bool,
+    /// The memory to use when returning an event to the caller.
+    pub memory: CoreItemRef<'a, kw::memory>,
+}
+
+impl<'a> Parse<'a> for CanonTaskWait<'a> {
+    fn parse(parser: Parser<'a>) -> Result<Self> {
+        parser.parse::<kw::task_wait>()?;
+        let mut l = parser.lookahead1();
+        let async_ = l.peek::<kw::r#async>()?;
+        if async_ {
+            parser.parse::<kw::r#async>()?;
+        }
```

Comment:
I think this can be simplified as `let async_ = parser.parse::<Option<kw::r#async>>()?.is_some();`

Reply:
(and in a few more places down below too)